frontend/libretro.c

```diff
@@ -303,6 +326,22 @@ static void vout_flip(const void *vram, int stride, int bgr24,
       }
    }
 
+   // Add the crosshair at a specified position (gunx, guny)
+   int crosshairSize_t = pl_rearmed_cbs.gpu_neon.enhancement_enable ? 4 : 2;
+   int crosshairSize_x = psx_w * (pl_rearmed_cbs.gpu_neon.enhancement_enable ? 2 : 1) / 40.0f;
+   int crosshairSize_y = psx_h * (pl_rearmed_cbs.gpu_neon.enhancement_enable ? 2 : 1) * (4.0f / 3.0f) / 40.0f;
+   for (port = 0; port < 2; port++) {
+      int gunx = (input_state_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_SCREEN_X) + 32767.0f) * dstride / 65534.0f;
+      int guny = (input_state_cb(port, RETRO_DEVICE_LIGHTGUN, 0, RETRO_DEVICE_ID_LIGHTGUN_SCREEN_Y) + 32767.0f) * vout_height / 65534.0f - vout_height;
+      if (in_enable_crosshair[port] > 0 && (in_type[port] == PSE_PAD_TYPE_GUNCON || in_type[port] == PSE_PAD_TYPE_GUN))
+      {
+         if (gunx == dstride)
+            addCrosshair(port, in_enable_crosshair[port], dest, dstride, gunx - 0.5f, guny, crosshairSize_t, crosshairSize_x, crosshairSize_y);
+	 else
+            addCrosshair(port, in_enable_crosshair[port], dest, dstride, gunx, guny, crosshairSize_t, crosshairSize_x, crosshairSize_y);
+      }
+   }
+
```

Comment:
Can you move this to a separate function and call it conditionally, name it `addCrosshairs` or something. This stuff is not used by vast majority of games, but you're needlessly calling input_state_cb() for every single one of them.

Reply:
I think this is done. I moved all of the size/position calculations to another function that is only called when both of these conditions are fulfilled:

- The Guncon or Konami Gun is being used in either port 1 or port 2
- The crosshair core option for that particular port is on

Reply:
Will squash again if this is okay.

Reply:
Yeah looks ok. Since you made `struct CrosshairInfo` you could pass it to `addCrosshair()` too but this is good enough as is.